Source/ARTLocalDevice.m

```diff
@@ -72,24 +71,18 @@ + (ARTLocalDevice *)load:(NSString *)clientId storage:(id<ARTDeviceStorage>)stor
     #endif
     device.push.recipient[@"transportType"] = ARTDevicePushTransportType;
 
-    NSString *deviceId = [storage objectForKey:ARTDeviceIdKey];
+    NSString *deviceId = [storage objectForKey:ARTDeviceIdKey] ?: @""; // PCD2, not nullable
```

Comment:
How does Android handle this — what value does `LocalDevice#id` return before RSH3a2b has happened?

Reply:
Android uses null(s).

Reply:
Yeah, I mean, there isn't a good answer to this situation; the spec is wrong. After this current PR is done, it might be worth you spending time making a PR for the spec changes necessary to fix it.

I think that the best option for now might be to ignore [RSH8b](https://sdk.ably.com/builds/ably/specification/main/features/#RSH8b) and to generate the new `LocalDevice.id` as soon as we clear the old one, so that the `LocalDevice` always has an `id`. I can't see any harm in doing this; what do you think?

Reply:
I think that temporarily not resetting `deviceId` in `resetDetails` and generate the new one for a new installation without persisting it might "do the trick". - https://github.com/ably/ably-cocoa/pull/1847/commits/094c9659b6c2318d132b62709a7e30083b78eadc and https://github.com/ably/ably-cocoa/pull/1847/commits/91183b9e4a1829e3d087cd23fdb9afca4fdc1410

Reply:
The aim of my suggestion was to make sure that the `LocalDevice` instance always has a non-nil `id`. In your implementation, it's possible for it to have a nil `id` after the device details have been cleared per RSH3g2a, no?

Reply:
But I don't think we want the `LocalDevice` to continue to have the same `id` after the device details have been cleared post-RSH3g2a, do we? Also, I think that someone looking at the code might not understand that we're relying on the `secret` having been cleared in order to generate a new device `id`.

I think it would better to, inside `resetDetails`, generate and persist a new `id`.

Reply:
Let's continue to generate `secret` at the point where the spec says that we should, yeah.

Reply:
What we have in this setup is `id` = `nil` until device activation which is inconsistent with what current property is (non nullable).

Reply:
That's a good point. So in that case we need to restore some of the previous behaviour - populate the `id` when the `LocalDevice` instance is first accessed, right?

Reply:
https://github.com/ably/ably-cocoa/pull/1847/commits/79e741b269e7ab7486292fc0acc09b5146c2badf

Still generating id/secret pair together, because once issue with nullable `id` resolved we can just delete extra `generateAndPersistPairOfDeviceIdAndSecret` calls, wdyt? Also it allows not to violate [RSH3a2b](https://sdk.ably.com/builds/ably/specification/main/features/#RSH3a2b) which states that those values should be generated as a pair.

Reply:
Yep, sounds good.